examples/gaussian_process/plot_gpr_on_structured_data.py

```diff
@@ -0,0 +1,137 @@
+"""
+==========================================================================
+Gaussian process regression and classification on discrete data structures
+==========================================================================
+
+This example illustrates the use of Gaussian processes to carry out
+regression and classification tasks on data that are not in fixed-length
+feature vector form. This is enabled through the use of kernel functions
+that can directly operate on discrete structures such as variable-length
+sequences, trees, and graphs.
+"""
+print(__doc__)
+
+import numpy as np
+import matplotlib.pyplot as plt
+from sklearn.gaussian_process.kernels import Kernel, Hyperparameter
+from sklearn.gaussian_process.kernels import StructureOrGenericKernelMixin
+from sklearn.gaussian_process import GaussianProcessRegressor
+from sklearn.gaussian_process import GaussianProcessClassifier
+from sklearn.base import clone
+
+
+class SequenceKernel(StructureOrGenericKernelMixin, Kernel):
+    '''
+    a mimimal (but valid) convolutional kernel for sequences of variable length
+    '''
+    def __init__(self,
+                 baseline_similarity=0.5,
+                 baseline_similarity_bounds=(1e-5, 1)):
+        self.baseline_similarity = baseline_similarity
+        self.baseline_similarity_bounds = baseline_similarity_bounds
+
+    @property
+    def hyperparameter_baseline_similarity(self):
+        return Hyperparameter("baseline_similarity",
+                              "numeric",
+                              self.baseline_similarity_bounds)
+
+    def _f(self, s1, s2):
+        return sum([1.0 if c1 == c2 else self.baseline_similarity
```

Comment:
I think a description would help. I think you can also generalise _f and _g essentially into a single function. Unless I'm much mistaken, `total = len(s1)*len(s2)`, then `_f(s1, s2) = total + _g(s1, s2) * (baseline_similarity - 1)`

Reply:
I can add a very short description that `_f` computes the kernel value while `_g` computes the kernel gradient. However, even though I agree that `_f` and `_g` can be simplified, I believe it is better not to merge them since that may obscure the interpretability. Another reason is that this kernel is not meanted to be used in any production code in the same way as the RBF or Matern kernel, but rather just as a mock/placeholder for demo purposes. I do plan to implement some more serious kernels for structured data and submit them through another PR. After that we can replace this minimal sequence kernel with those new ones.

Reply:
Mmm... Well for now the difference in line breaks obscures their similarity which is also not good!

Reply:
So how about making the line-breaking style consistent like this?
```
    def _f(self, s1, s2):
        '''
        kernel value between a pair of sequences
        '''
        return sum([1.0 if c1 == c2 else self.baseline_similarity
                   for c1 in s1
                   for c2 in s2])

    def _g(self, s1, s2):
        '''
        kernel derivative between a pair of sequences
        '''
        return sum([0.0 if c1 == c2 else 1.0
                    for c1 in s1
                    for c2 in s2])
```
Meanwhile, as I mentioned, I am expecting to replace this kernel soon via another PR with a more serious one, that will be defined alongside all the RBF, Matern, DotProduct kernels etc. So we probably don't need to worry to much about the current one anyway.

Reply:
Except in an example, I don't think you should expect that collections of structured kernels will be merged into scikit-learn. We have explicitly left support for most non-vectorial data out of scope, to reduce the maintenance burden.

Yes, consistent formatting makes it easier to read the difference.

Reply:
Sad to hear this decision on leaving out non-vector data... I understand that supporting these data across the entire project could be challenge. However, since kernels are particularly suitable for processing structured data by hiding details of data form from higher level ML algorithms, would it be reasonble to support non-vector data just for GPR/GPC/etc.? This would be parallel to the fact that some of the GP kernels are mathematically equivalent to members of `pairwise.metric`, yet are tailored just for GP learning.

Reply:
Part of the problem is having to design standard input representations. For genetic sequences this is relatively straightforward and standardised, but not so for trees, heterogeneous time series, etc. I would much rather see that in an externally maintained package that we can mention in our documentation than for it to be something for the core devs to consider, authorise, and ultimately maintain. I hope you can trust that even within the basic scope of vectorial data there is enough and diverse code to maintain here.

Reply:
I see. Thanks for the explanation!